app/main.py

```diff
@@ -15,6 +18,103 @@
+
+    # TODO: this should not be done in the future! probably not even in debug mode.
+    logger.info(f"request.headers: {request.headers}")
+
+    # check env to see if auth checking should be performed
+    # TODO: move to config, properly boolean-ize
+    if os.getenv("OLS_AUTH_CHECK") == "False":
+        logger.info(f"auth checks disabled, skipping")
+        return await call_next(request)
+
+    # ignore auth for default routes and docs
+    # TODO: is there a way to programmatically determine which paths/routes we want to skip auth on?
+    if request.url.path in [
+        "/",
+        "/readyz",
+        "/healthz",
+        "/status",
+        "/docs",
+        "/openapi.json",
+    ]:
+        logger.info(f"ignoring default route for auth check")
+        return await call_next(request)
+
+    # not a default, so check if the user is authorized to use ols via the k8s auth
+
+    # if there's no authorization header, fail with noauth http error
+    if "authorization" not in request.headers:
+        logger.info(f"no auth header found")
+        return JSONResponse(
+            status_code=requests.codes.unauthorized,
+            content={"message": "no auth header found"},
+        )
+
+    logger.info(f"auth header found, checking k8s auth")
+
+    if not k8s_auth(request.headers["authorization"]):
+        return JSONResponse(
+            status_code=requests.codes.unauthorized,
+            content={"message": "not authorized to use ols service"},
+        )
+
+    # if we made it this far, they are authorized, so continue
+    logger.info(f"passed authorization check")
+    return await call_next(request)
+
+
+def k8s_auth(authorization) -> bool:
+    configuration = kubernetes.client.Configuration()
+
+    # strip the token from the header
+    # Configure API key authorization: BearerToken
+    configuration.api_key["authorization"] = authorization.split("Bearer ")[1]
+
+    # Uncomment below to setup prefix (e.g. Bearer) for API key, if needed
+    configuration.api_key_prefix["authorization"] = "Bearer"
+
+    # Defining host is optional and default to http://localhost
+    configuration.host = os.getenv("K8S_CLUSTER_API")
+
+    # TODO: make configurable
+    # cluster is using a self-signed cert -- would need to inject CA information
+    configuration.verify_ssl = False
+
+    # create a k8s client using the supplied authentication information
+    k8s_client = kubernetes.client.ApiClient(configuration)
+
+    # create a self subject access review to determine if the user can access the lightspeed service
+    # TODO: this is currently checking a non-resource attribute path of `/ols` which is totally generic.
+    #       in the future we might want to restrict users to specific ols endpoints even and make this much
+    #       fancier
+    jsonstr = '{"apiVersion":"authorization.k8s.io/v1","kind":"SelfSubjectAccessReview","spec":{"nonResourceAttributes":{"path":"/ols","verb":"get"}}}'
```

Comment:
rather than us passing the user's token on to k8s, i think we should be determining the user's identity (from the token) and then doing a subject access review in which we ask k8s "can user X do Y"?

but this is an area where we probably need to get some advice from the auth team.

Reply:
My question is simply: Why?

The user passed us their token. Why would we first determine who they are, and then ask if they can do a thing? We already are them via their key. That's why this is a SelfSubjectAccessReview -- because we are that user (via their key) asking if *I* can do the thing.

Reversing the discussion - once we determine the user can do the thing, then what?  The purpose of using their key was so that later/in the future, we can interact with the API at their permission level to do things. For example, we use their API key to decide if they can use OLS, and then, if they can, we use their API key to maybe modify an object in the cluster. 

The user had access to OLS, but did not have write/edit access on the resource in question, so they were blocked.

If we didn't use their key, we'd end up having to do accessreview requests for pretty much everything, which seems wasteful/inefficient.

:shrug:

Reply:
> My question is simply: Why?

Passing the user token around feels like an unnecessary risk.  The token is very powerful.

The ideal flow in my mind would be more like a traditional oauth flow in which OLS just gets a token that only has enough scope/permission to verify the user's identity, but can't do anything else.

but more fundamentally us defining auth flows w/ talking to the auth experts worries me.

Reply:
I agree that we should be concerned about what we could do with the user's token.

I agree that we should talk to someone about how we could get more granular, but understanding how granular we need to get requires us to have a clear picture of what we want to be able to do on behalf of the user. Right now the scope is "whatever the user could do, OLS should be able to do on their behalf."

Remember, at some point we want the user to be able to do things like "can you tell me which pods are crash looping?"

This requires a lookup. It sounds like you are saying you don't want OLS to do that. Or, if you want OLS to do that, you want a system that _temporarily_ gets enough permissions to do _only_ that.

?

Reply:
> This requires a lookup. It sounds like you are saying you don't want OLS to do that. Or, if you want OLS to do that, you want a system that temporarily gets enough permissions to do only that.

I definitely don't want to give it permission to do that until we are committed to it doing that and have fully thought through the approach.  So i'm focused on "what's the right way to do the thing we know we need to do now" and trying to avoid  "let's open a potential sudo hole that will allow us to do things we might want to do in the future"

And if someone can come back and say "yeah, using the user token for this is fine", then i'm ok with it.  I believe it is what the console does, so it's not unprecedented.  I just want us to do the exercise of confirming the approach.

Reply:
ACK